src/MSBuild.Sdk.SqlProj.Aspire/PublishSqlProjectLifecycleHook.cs

```diff
@@ -1,67 +1,176 @@
         _resourceNotificationService = resourceNotificationService ?? throw new ArgumentNullException(nameof(resourceNotificationService));
     }
 
-    public async Task AfterResourcesCreatedAsync(DistributedApplicationModel application, CancellationToken cancellationToken)
+    public Task BeforeStartAsync(DistributedApplicationModel application, CancellationToken cancellationToken = default)
     {
-        foreach (var sqlProject in application.Resources.OfType<SqlProjectResource>())
+        foreach (var resource in application.Resources.OfType<SqlProjectResource>())
         {
-            var logger = _resourceLoggerService.GetLogger(sqlProject);
+            StartDacpacDeployment(resource, application, _tokenSource.Token);
+        }
 
-            var dacpacPath = sqlProject.GetDacpacPath();
-            if (!File.Exists(dacpacPath))
-            {
-                logger.LogError("SQL Server Database project package not found at path {DacpacPath}.", dacpacPath);
-                await _resourceNotificationService.PublishUpdateAsync(sqlProject,
-                    state => state with { State = new ResourceStateSnapshot("Failed", KnownResourceStateStyles.Error) });
-                continue;
-            }
+        return Task.CompletedTask;
+    }
 
-            var targetDatabaseResourceName = sqlProject.Annotations.OfType<TargetDatabaseResourceAnnotation>().Single().TargetDatabaseResourceName;
-            var targetDatabaseResource = application.Resources.OfType<SqlServerDatabaseResource>().Single(r => r.Name == targetDatabaseResourceName);
-            var connectionString = await targetDatabaseResource.ConnectionStringExpression.GetValueAsync(cancellationToken);
-            if (connectionString == null)
-            {
-                logger.LogError("Failed to retrieve connection string for target database {TargetDatabaseResourceName}.", targetDatabaseResourceName);
-                await _resourceNotificationService.PublishUpdateAsync(sqlProject,
-                    state => state with { State = new ResourceStateSnapshot("Failed", KnownResourceStateStyles.Error) });
-                continue;
-            }
+    private void StartDacpacDeployment(SqlProjectResource sqlProject, DistributedApplicationModel application, CancellationToken cancellationToken)
+    {
+        var logger = _resourceLoggerService.GetLogger(sqlProject);
 
-            await _resourceNotificationService.PublishUpdateAsync(sqlProject,
-                state => state with { State = new ResourceStateSnapshot("Publishing", KnownResourceStateStyles.Info) });
+        _ = Task.Run(async () =>
+        {
+            int retryCount = 10;
+            int delayInMillisecodns = 1000;
 
-            try
+            for (int i = 0; i < retryCount; i++)
             {
-                _deployer.Deploy(dacpacPath, connectionString, targetDatabaseResource.DatabaseName, logger, cancellationToken);
+                try
```

Comment:
This feels like a lot of code now for one method. Perhaps we could break this out a little bit? Maybe a WaitForServerReady and Publish method or something?